trie/committer.go

```diff
@@ -0,0 +1,284 @@
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+	"golang.org/x/crypto/sha3"
+)
+
+// LeafChanSize is the size of the leafCh. It's a pretty arbitrary number, to allow
+// some paralellism but not incur too much memory overhead.
+const LeafChanSize = 200
+
+// Leaf represents a trie leaf value
+type Leaf struct {
+	size   int         // size of the rlp data (estimate)
+	hash   common.Hash // hash of rlp data
+	node   node        // the node to commit
+	vnodes bool        // set to true if the node (possibly) contains a valueNode
+}
+
+// committer is a type used for the trie Commit operation. A committer has some
+// internal preallocated temp space, and also a callback that is invoked when
+// leaves are committed. The leafs are passed through the `leafCh`,  to allow
+// some level of paralellism.
+// By 'some level' of parallelism, it's still the case that all leaves will be
+// processed sequentially - onleaf will never be called in parallel or out of order.
+type committer struct {
+	tmp sliceBuffer
+	sha keccakState
+
+	onleaf LeafCallback
+	leafCh chan *Leaf
+}
+
+// committers live in a global db.
+var committerPool = sync.Pool{
+	New: func() interface{} {
+		return &committer{
+			tmp: make(sliceBuffer, 0, 550), // cap is as large as a full fullNode.
+			sha: sha3.NewLegacyKeccak256().(keccakState),
+		}
+	},
+}
+
+// newCommitter creates a new committer or picks one from the pool, and
+// initializes the leafCh, if needed.
+// In case no onleaf-callback is provided, the committer does not
+// use a channel-based commit, but inlined.
+// Typically, the account trie is committed with a channel-based leaf-commit,
+// whereas storage tries are committed 'inline'.
+func newCommitter(onleaf LeafCallback) *committer {
+	h := committerPool.Get().(*committer)
+	h.onleaf = onleaf
+	if onleaf != nil {
+		h.leafCh = make(chan *Leaf, LeafChanSize)
+	}
+	return h
+}
+
+func returnCommitterToPool(h *committer) {
+	h.onleaf = nil
+	h.leafCh = nil
+	committerPool.Put(h)
+}
+
+// commitNeeded returns 'false' if the given node is already in sync with db
+func (h *committer) commitNeeded(n node) bool {
+	hash, dirty := n.cache()
+	return hash == nil || dirty
+}
+
+// hash collapses a node down into a hash node, also returning a copy of the
+// original node initialized with the computed hash to replace the original one.
+func (h *committer) commit(n node, db *Database, force bool) (node, error) {
+	// If we're not storing the node, just hashing, use available cached data
+	hash, dirty := n.cache()
+	if hash != nil && !dirty {
+		return hash, nil
+	}
+	if db == nil {
+		return nil, errors.New("no db provided")
```

Comment:
Wondering about this. Should we error or panic? If it's a comitter, should it allow nil db? Isn't that a programming error?

Reply:
It also feels a bit weird to explicitly check this for every single trie node. Shoudln't this be checked way way above and here just assumed it's correct?

Reply:
If db is `nil`, it means that the `trie.db` is nil. I can't say if that's ever done on purpose, e.g. in tests.